cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -89,16 +93,17 @@ public void init(HandlerContext context) {
   }
 
   /**
-   * Sends control commands to the client.
+   * Sends control commands to the client. Also receives program status updates from the client.
    */
-  @GET
+  @PUT
   @Path("/tethering/controlchannels/{peer}")
   public void connectControlChannel(FullHttpRequest request, HttpResponder responder, @PathParam("peer") String peer,
                                     @QueryParam("messageId") String messageId)
     throws IOException, NotImplementedException, PeerNotFoundException, ForbiddenException, BadRequestException {
     checkTetheringServerEnabled();
     store.updatePeerTimestamp(peer);
     TetheringStatus tetheringStatus = store.getPeer(peer).getTetheringStatus();
+    processProgramUpdates(request.content().toString(StandardCharsets.UTF_8));
```

Comment:
It's better to convert the request body in a streaming fashion instead of converting to a String and then converting to an object. Something like:

```

    try (Reader reader = new InputStreamReader(new ByteBufInputStream(request.content()), StandardCharsets.UTF_8)) {
      propertyRequests = GSON.fromJson(reader, TetheringProgramUpdateRequest.class);
    } catch (JsonSyntaxException e) {
      throw new BadRequestException("Unable to parse request: " + e.getMessage(), e);
    }
```